pennylane/transforms/split_non_commuting.py

```diff
@@ -11,39 +11,45 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+
 """
-Contains the tape transform that splits non-commuting terms
+Contains the tape transform that splits a tape into tapes measuring commuting observables.
 """
-from functools import reduce
 
-# pylint: disable=protected-access
-from typing import Callable, Sequence
+from functools import partial
+from typing import Callable, Dict, List, Sequence, Tuple
 
 import pennylane as qml
+from pennylane.measurements import ExpectationMP, MeasurementProcess, Shots, StateMP
+from pennylane.ops import Hamiltonian, LinearCombination, Prod, SProd, Sum
 from pennylane.transforms import transform
-
-
-def null_postprocessing(results):
-    """A postprocesing function returned by a transform that only converts the batch of results
-    into a result for a single ``QuantumTape``.
-    """
-    return results[0]
+from pennylane.typing import Result, ResultBatch
 
 
 @transform
-def split_non_commuting(tape: qml.tape.QuantumTape) -> (Sequence[qml.tape.QuantumTape], Callable):
-    r"""
-    Splits a qnode measuring non-commuting observables into groups of commuting observables.
+def split_non_commuting(
+    tape: qml.tape.QuantumScript,
+    group: bool = True,
+    grouping_strategy: str = None,
+) -> (Sequence[qml.tape.QuantumTape], Callable):
+    r"""Splits a tape into tapes measuring groups of commuting observables.
 
     Args:
-        tape (QNode or QuantumTape or Callable): A circuit that contains a list of
-            non-commuting observables to measure.
+        tape (~pennylane.tape.QuantumScript): The tape to be split.
+        group (bool): Whether to compute disjoint groups of commuting observables,
+            leading to fewer tapes. If ``group=False``, one tape will be generated for
+            each observable to be measured.
+        grouping_strategy (str): The grouping strategy to be used. If "naive",
+            grouping will be performed based on overlapping wires; if "pauli",
+            grouping is computed using :func:`~pennylane.pauli.group_observables`.
+            If ``group=False``, this argument is ignored.
```

Comment:
Maybe combine this into a single parameter? Let `group` be `bool or str`, `group=False` for no grouping, `group=True` for yes grouping, and `group="naive"` for yes and use naive grouping?

Reply:
Nice!

Though maybe just have `grouping_strategy=None` (default, no grouping), with options:
- `"naive"`
- `"pauli"`
I guess we had discussion a few days back on what the default should be? :thinking:

Reply:
I don't think the default was ever no grouping. Previously `split_non_commuting` always does pauli grouping, `hamiltonian_expand` does pauli grouping by default unless `group=False`, and `sum_expand` does naive grouping by default unless `group=False`.

Reply:
> I don't think the default was ever no grouping. Previously `split_non_commuting` always does pauli grouping, `hamiltonian_expand` does pauli grouping by default unless `group=False`, and `sum_expand` does naive grouping by default unless `group=False`.

Thanks for the summary!

> Instead of setting some arbitrary threshold, I've made the assumption (similar to our current logic in Device.batch_transform) that the problematically large hamiltonians typically come in the form of an Hamiltonian or LinearCombination, so if we see any of these two in the list of observables, we fall back to naive grouping.

What is our current precedent? if a user has a `qml.expval(H)` return in their QNode, which transform are we currently using? `hamiltonian_expand`?

Reply:
| Scenario | Current Behaviour | New Behaviour |
| --------- | ------------------ | --------------- |
| Single measurement of a `Hamiltonian`/`LinearCombination` | `hamiltonian_expand` (pauli grouping) | pauli grouping |
| Single measurement of a `Sum` | `sum_expand` (naive grouping) | pauli grouping (completes sc-59046) |
| Multiple measurements containing an `expval` of a `Hamiltonian` /`LinearCombination`| `sum_expand` (naive grouping) | naive grouping |
| Multiple measurements containing an `expval` of `Sum`, `Prod`, or `SProd` | `sum_expand` (naive grouping) | pauli grouping |
| Multiple measurements only of single term observables | `split_non_commuting` (pauli grouping) | pauli grouping |

Reply:
I believe the idea is that, since pauli grouping can be expensive to compute, and if we were to stick this transform into device preprocess, it is performed for every circuit execution, so to save some time, we try to avoid doing pauli grouping, unless:
1. There is only a single measurement of a `Sum`/`Hamiltonian`, in which case the grouping is computed once and cached in the `Hamiltonian`/`Sum` object and reused for every execution
2. There are multiple measurements but none of them is a `Hamiltonian`/`LinearCombination`, in which case we assume that there probably aren't that many of them, so we can afford to recompute the grouping for every tape execution.

This is under the assumption that if there is a huge multi-term observable, it's probably defined as: `qml.Hamiltonian([...], [...])` instead of `obs1 + obs2 + obs3 + ...`. I originally had the idea of setting a threshold based on the total number of observables in the tape, but it wasn't clear what that arbitrary threshold would be.

Admittedly the current solution doesn't make perfect sense, but it's most consistent with the current decision logic. Ultimately we want to figure out a way to cache that grouping information somewhere globally and retrieve it reliably, then we will be able to use pauli grouping for all scenarios.

Reply:
For some historical context: when @mariaschuld and I originally added support for `return qml.expval(H)`, we made the decision to only perform QWC grouping _if the user had already pre-computed the grouping indices_ and they were available via `H.grouping_indices` (or similar, can't remember the exact attribute name).

This was a compromise to avoid expensive grouping computations everywhere, and an attempt to try and give the user control over whether grouping happens or not.

That would still be my preference if possible -- make this something that user can specify, rather than us attempting to use heuristics.

Reply:
@josh146 The current default behaviour of `hamiltonian_expand` before this PR is `group=True`, which means that `H.compute_grouping()` is called if pre-computed grouping indices are not present. The user is able to manually turn it off by setting `group=False`. I'm not sure when it became this way but it's been like this for a couple of releases. With this PR the user is still able to specify `grouping_strategy=None` if they do not want grouping to be computed, but if `H.grouping_indices` is present, it will be used regardless, same as before.

Reply:
Another thing to mention is that while we discuss grouping indices being expensive, we are using an extremely inefficient method of computing grouping indices. 3 weeks of work would reduce the performance cost by orders of magnitude and eliminate the classical cost concerns for the immediate future.

`grouping_indices` are also fairly limiting. It's limited to individual operators, forces each to term to only exist in only one measurement basis, and has to be recalculated even if only the coefficients change.

Reply:
Thanks for the update! @albi3ro that's really good to know -- might be worth making a shortcut story with the details, and adding it to the backlog. We can then potentially pull it in in Q3